### DISCUSSION TITLE
the R function mclapply() in the parallel package doesn't seem to work when rendering a quarto document

### DISCUSSION THREAD

**Discussion by paulgribble** [quarto-dev/quarto-cli#8734] [Q&A]
### Description

The following R code when placed within a quarto document and rendered with `quarto preview` only uses 1 core even when instructed to use 16 cores (my machine has 16 cores).

The identical code in an RStudio or R session does use all 16 cores.

Is this a limitation somehow of quarto?

```{r}
library(parallel)
work <- function(n) {
  ng <- 3
  df <- tibble(x=c(rnorm(ng,100,15), rnorm(ng,100,15), rnorm(ng,100,15)), 
               g=factor(c(rep("group1",ng), rep("group2",ng), rep("group3",ng))))
  summary(aov(x~g, data=df))[[1]]$"F value"[1]
}
nsims = 10000
ng = 3
n = rep(10, nsims)
results <- mclapply(n, work, mc.cores=16)
fvals <- rep(0, nsims)
for (i in 1:nsims) {
  fvals[i] <- results[[i]]  
}
fvals <- tibble(fvals)
```

**Comment by mcanouil** [comment]
Did you try within an Rmarkdown document?
If you observed the same behaviour, it means it's not Quarto.

````rmd
---
output: html_document
---

```{r}
library(parallel)
work <- function(n) {
  ng <- 3
  df <- tibble(x=c(rnorm(ng,100,15), rnorm(ng,100,15), rnorm(ng,100,15)), 
               g=factor(c(rep("group1",ng), rep("group2",ng), rep("group3",ng))))
  summary(aov(x~g, data=df))[[1]]$"F value"[1]
}
nsims = 10000
ng = 3
n = rep(10, nsims)
results <- mclapply(n, work, mc.cores=16)
fvals <- rep(0, nsims)
for (i in 1:nsims) {
  fvals[i] <- results[[i]]  
}
fvals <- tibble(fvals)
```
````

**Reply by cscheid** [reply]
We shouldn't expect people to know rmarkdown in other to diagnose things about Quarto.

**Reply by mcanouil** [reply]
Fair point. (although very unlikely an R user today does not know anything about rmarkdown)

**Comment by cscheid** [comment]
> even when instructed to use 16 cores

How did you do this instruction? Quarto has no special parallel handling; I think what's happening is that your configuration is not getting communicated to the library.

**Reply by paulgribble** [reply]
OK I tried it in RMarkdown and it successfully uses multiple cores. The same code in a quarto document when rendered or previewed does not.

even when I hard code the number of cores by calling `results <- mclapply(n, work, mc.cores=16)` instead of using the detected number of cores, when rendered using Quarto it still only uses 1 core.

I guess this means it is in fact somehow Quarto's fault. Strange

Here is the code

```{r}
library(tidyverse)
library(parallel)
work <- function(n) {
  ng <- 3
  df <- tibble(x=c(rnorm(ng,100,15), rnorm(ng,100,15), rnorm(ng,100,15)), 
               g=factor(c(rep("group1",ng), rep("group2",ng), rep("group3",ng))))
  summary(aov(x~g, data=df))[[1]]$"F value"[1]
}
nsims = 10000
ng = 3
n = rep(10, nsims)
numCores <- detectCores()
results <- mclapply(n, work, mc.cores=numCores)
fvals <- rep(0, nsims)
for (i in 1:nsims) {
  fvals[i] <- results[[i]]  
}
fvals <- tibble(fvals)
ggplot(fvals, aes(x=fvals)) + geom_histogram(bins=50)
```

**Reply by mcanouil** [reply]
Could you try something simple? As I am not able to reproduce your described behaviour.
Your 10,000 processes are very small. By the way, not a good idea to fork parallel like that on small tasks. You spend more time creating the forked R processes and retrieving the results than computing the results.


````qmd
---
format: html
---

```{r}
library(parallel)
n_cores <- 16
system.time({
  results <- mclapply(
    X = seq_len(n_cores),
    FUN = function(n) {
      Sys.sleep(60)
    },
    mc.cores = n_cores
  )
})
```
````

With the above code, if it runs in parallel, the elapsed time will be 60 seconds, otherwise 16 times 60 seconds.

| 10 Cores | 1 Core |
|--------|--------|
| <img width="1624" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/a3621600-41a7-4284-baf7-951fd0524509"> | <img width="1624" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/30cf7b4d-115f-4925-853b-19071059db90"> |

**Comment by paulgribble** [comment]
I `quarto preview`'d in VSCode with your code above and it was indeed 60 sec so it was using all cores. Strange.

thanks for the help I think I know what to do

<img width="1488" alt="Screenshot 2024-02-14 at 11 22 15 AM" src="https://github.com/quarto-dev/quarto-cli/assets/2848/95499440-0a1a-4c13-9ffc-e7af189a609d">
